pulsar/reader_partition.go

```diff
@@ -0,0 +1,142 @@
+package pulsar
+
+import (
+	"sync"
+	"time"
+)
+
+func (r *reader) internalTopicReadToPartitions() error {
+	partitions, err := r.client.TopicPartitions(r.topic)
+	if err != nil {
+		return err
+	}
+
+	oldNumPartitions, newNumPartitions := 0, len(partitions)
+
+	r.Lock()
+	defer r.Unlock()
+
+	oldReaders, oldNumPartitions := r.consumers, len(r.consumers)
+	if oldReaders != nil {
+		if oldNumPartitions == newNumPartitions {
+			r.log.Debug("Number of partitions in topic has not changed")
+			return nil
+		}
+
+		r.log.WithField("old_partitions", oldNumPartitions).
+			WithField("new_partitions", newNumPartitions).
+			Info("Changed number of partitions in topic")
+	}
+
+	r.consumers = make([]*partitionConsumer, newNumPartitions)
+
+	// When for some reason (eg: forced deletion of sub partition) causes oldNumPartitions> newNumPartitions,
+	// we need to rebuild the cache of new consumers, otherwise the array will be out of bounds.
+	if oldReaders != nil && oldNumPartitions < newNumPartitions {
+		// Copy over the existing consumer instances
+		for i := 0; i < oldNumPartitions; i++ {
+			r.consumers[i] = oldReaders[i]
+		}
+	}
+
+	type ConsumerError struct {
+		err       error
+		partition int
+		consumer  *partitionConsumer
+	}
+
+	subscriptionName := r.options.SubscriptionRolePrefix
+	if subscriptionName == "" {
+		subscriptionName = "reader"
```

Comment:
In here, maybe we can define a string constant

```
const ReaderSubNamePrefix = "reader"
```

Reply:
ok